added private_domains functionality

jaredmcqueen opened the pull request:
I changed the default TLD parsing behavior to ignore private domains.  If users want private domains, they can set a new config parameter called private_domains to true.

Tests were added, as was documentation.  I haven't tested how the inline documentation will render.

I also got rid of the default example filter text found in the inline documentation.

jordansissel commented on lib/logstash/filters/tld.rb:
```diff
@@ -2,10 +2,32 @@
 
-# This example filter will replace the contents of the default 
-# message field with whatever you specify in the configuration.
+# This filter is a domain name parser based on the https://publicsuffix.org/[Public Suffix List]
```
docs live here: https://github.com/logstash-plugins/logstash-filter-tld/blob/master/docs/index.asciidoc

(This is a recent change)

jaredmcqueen replied:
ok, not familiar with modifying asciidocs directly - i thought they were auto-generated based on inline comments.  I can change these as well.

jordansissel commented on lib/logstash/filters/tld.rb:
```diff
@@ -18,17 +40,21 @@ class LogStash::Filters::Tld < LogStash::Filters::Base
-    # Add instance variables 
+    # Add instance variables
     require 'public_suffix'
+    PublicSuffix::List.private_domains = @private_domains
```
This is an interesting setting. It appears to be a global setting which means the last tld plugin to initialize will win. This impacts users who may have multiple pipelines (coming in Logstash 6.0, iirc) where two pipelines may use a tld filter.

Thoughts?

Maybe we have to modify the upstream library to not use globals.

jaredmcqueen replied:
Good catch.  I'll modify according to how the author does it here: https://github.com/weppos/publicsuffix-ruby#private-domains

jordansissel commented on lib/logstash/filters/tld.rb:
```diff
@@ -18,17 +40,21 @@ class LogStash::Filters::Tld < LogStash::Filters::Base
   config :target, :validate => :string, :default => "tld"
 
+  # Allows private (non-ICANN) domain parsing
+  config :private_domains, :validate => :boolean, :default => false
```
I confess to not understanding what this setting enables. From reading the tests, it considers "s3.amazonaws.com" to be a tld? I'm confused.

To ask a more pointed question, how will a user know what this setting does?

jaredmcqueen replied:
The vast majority of customers / users just want public TLD parsing - to see the .com, .net, etc.  They are always surprised (and confused) when they see a TLD like s3.amazonaws.com because they expect .com, not the whole thing.  

Search "===begin private domains==="  in the list at https://publicsuffix.org/list/public_suffix_list.dat and take a look for yourself.

for all Machine Learning, cyber security, anomaly detection use cases, customers just want public -- NOT private

jordansissel commented:
Thanks for helping improve this plugin :)

I've left some comments on the PR.

jaredmcqueen commented:
@jordansissel I updated the filter to disable private domains by default.  Users can enable it using the config `ignore_private`.  I updated the asciidocs and tests.

jordansissel commented:
I have this PR on my todo list for this week.

jordansissel commented on Gemfile.lock:
```diff
@@ -0,0 +1,118 @@
+PATH
```
no need to add Gemfile.lock to the git repo. Can you remove this file?

jaredmcqueen replied:
yes, done.

jordansissel replied:
This file is still here, according to github?

jaredmcqueen replied:
oops.. deleted it.  sorry about that.

jordansissel commented:
overall LGTM. 

I left one comment asking that Gemfile.lock file be removed (or tell me a story why you added it). Once this is resolved, I will merge.

Thank you for working on this :)

jordansissel commented:
This PR is failing on Logstash 5.5 branch: 

> public_suffix-2.0.5 requires ruby version >= 2.0, which is incompatible with the current version, ruby 1.9.3p551 (jruby 1.7.25)

jaredmcqueen commented:
ok I'm back.  It looks like since the earlier versions of logstash use 1.9.3, we are limited to using a *really* old version of public_suffix (1.4.6).   In 1.4.6, the private / public parsing is all or nothing.  You cannot choose to parse private/public on a per-use of the filter basis.

Still though, the default really should be to ignore the private domains.  Submitting my changes, and will wait for Logstash 6.0 branch to use the new and improved public_suffix version that has more control over domain parsing.

jaredmcqueen commented:
@jordansissel standing by for your input on this request, let me know if there's anything else you need

jordansissel commented:
LGTM

jaredmcqueen commented:
any word if this will get pulled?  Asking on behalf of a few customers.

jaredmcqueen commented:
any update on this?

jksecurity commented:
I am also interested in using the updated version of this plugin. Any ETA on when this gets merged?
